### ISSUE TITLE
Identity of Concourse user should be threaded through to other components in a single-identity-server orchestration environment

### ISSUE THREAD

**Issue by jraqula** [concourse/concourse#2274]
# Feature Request

## What challenge are you facing?

I find that I use Concourse to manage environment state via a variety of largely output-only resources such as the BOSH deployment resource and the CF resource. Though I have misgivings about this flow design, it is [somewhat well discussed already](https://github.com/concourse/concourse/issues/413#issuecomment-233076618) and seems to be an increasingly supported strategy.

As part of this, I use github and UAA for our authentication into our concourse. I would like to be able to use UAA only, which manages authentication for some of my BOSH/CF work as well.

A kink in the setup at the moment is that we would like to require specific users to authorize the deploy actions for audit trail & security reasons, but when triggering the build, Concourse acts as its own pre-authenticated client.

In other words, one acceptance criteria for this might be that my concourse does not contain deploy-authorized client credentials yet I am still able to trigger BOSH deploys via press-the-+-button.

Broadly, I am really unsure how to design a flow around this which is not a transgression against the design principles of Concourse. For example, a workaround where I obtain a hardcopy of a fresh access token and paste it into a git repo is obvious. But that is unappealing for a plethora of reasons.
 However, it has got me thinking on a new angle about identity management & environmental setup.

## A Modest Proposal

The one thought that comes to mind here that seems stable, BUT which is a too snowflakey for my taste, is that I might elect to give builds access to my current access token when I do `trigger-build`. this would be useful only when the identity server is shared between concourse and some third party entity being managed, such as BOSH. This makes assumptions I don't like to make, and dirties up the separation of concerns, but may provide the functionality in what I hope is a pretty common use case.

Very interested to hear thoughts here and have a conversation :) Thanks!

**Comment by marco-m** [comment]
FYI, there are a lot of open tickets related to what you are asking, see for example 
- https://github.com/concourse/concourse/labels/epic%2Fusers
- https://github.com/concourse/concourse/labels/auth
- https://github.com/concourse/concourse/labels/multi-tenancy

Regarding 

> For example, a workaround where I obtain a hardcopy of a fresh access token and paste it into a git repo is obvious. But that is unappealing for a plethora of reasons.

I can propose an alternative workaround that has the advantage of not needing any intermediary "paste in a git repo" and is not so much a workaround actually:

- Use a task param (https://concourse-ci.org/tasks.html#task-params). A task param is made available to the task as an environment variable.
- Then, pass the access token with fly execute, see section "Providing values for params" in https://concourse-ci.org/running-tasks.html#fly-execute.

**Comment by jraqula** [comment]
Those references are relevant I agree, and I expect this issue might be tagged with the `auth` label here. And that `fly execute` strategy is a good idea, however doesn't work with the `+` button via Web UI if one were interested in doing that. Also, can you describe how to use it with side effects you are trying to accomplish with `put` steps? Those are my primary use case.

As far as my searching shows me, this discussion is unique among the ones you link to because it pertains to relating your concourse authentication identity to side effects you want to accomplish with tasks/puts. As implied in my closing comment above, this may interact interestingly with changes to the authentication mechanism in Concourse, and could be considered an extra feature in the course of that track of work.

From what I can tell however, it  mandates a change to concourse-bespoke logic & features. Even if authentication is done via `dex` or any other provider, Concourse would in this case need to have a first-class system behavior to make authentication which is done at a user level to be made available in a dynamic way to steps, which are statically defined.

**Comment by jama-pivotal** [comment]
Indeed, the problem space sounds quite familiar to the folks I've spoken to in the past. In fact, the workaround I've seen is extreme and quite burdensome: breaking up pipelines completely in two; one half for the working team and the other half to the authorized team who is "trusted" to press (+)

Even with the addition of `dex`, our short term accomplishments merely allow you to be identified as an individual. There will be no RBAC right out the gate (#1317) and we're working towards a more robust logging model for auditing relevant user actions (no issue open..yet!)

I'm trying to imagine a world where a user can UAA authenticate into Concourse via `dex`, and we have RBAC in place that entitles the trigger action (+) on a role, _and_ we can reliably output logs of what action was executed by a user. Would that be enough? Is that too constrictive? Would we still need to flow the auth data to the action in that case? 🤔

**Comment by marco-m** [comment]
@jraqula 

> Also, can you describe how to use it with side effects you are trying to accomplish with put steps?

You made me realize that my suggestion is wrong. You can pass parameters with `fly execute`, but they would be limited to a specific task, they would not "propagate" to a resource part of the same pipeline that contained the task. Too bad!

The other workaround I can think of is the following, which has the advantage, compared to the "split in 2 pipelines with different teams", to stay in the same pipeline. I tested it and it works.

In the following pipeline I am using the S3 resource instead than the BOSH resource because I don't have a BOSH environment, but the concept is the same.

The main idea is to use a secret store and inject the needed parameter only in the limited time window when a deploy is needed.

Consider the pipeline:

```
resources:

  - name: destination.s3
    type: s3
    source:
      regexp: artifact-(.*)
      bucket: channel
      endpoint: ((minio-endpoint))
      access_key_id: ((s3-access-key-id))
      secret_access_key: ((s3-secret-access-key))

jobs:

  # - name: the-real-build-job
  #   ...

  - name: deploy
    plan:
      - task: deploy
        config:
          platform: linux
          image_resource:
            type: docker-image
            source: {repository: alpine}
          outputs:
            - name: out
          run:
            path: sh
            args:
              - -c
              - |
                set -o errexit
                touch out/artifact-$(date +'%Y%m%d%H%M%S')
      - put: destination.s3
        params:
          file: out/artifact-*
```

We have 3 parameters: `((minio-endpoint))`, `((s3-access-key-id))`, `((s3-secret-access-key))`. We store in Vault the first 2, and we leave out the 3rd, `((s3-secret-access-key))`, which represents in your case the BOSH deploy secret.

I show the interaction on the commandline, but this approach supports also triggering from the web UI.

First trigger, will fail because missing parameter. This emulates somebody without credentials:

```
fly -t vm trigger-job -j deploy/deploy -w
started deploy/deploy #3

initializing
running sh -c set -o errexit
touch out/artifact-$(date +'%Y%m%d%H%M%S')

Expected to find variables: s3-secret-access-key
errored
```

We are now ready to deploy. We temporarily inject the secret in Vault:

```
vault kv put /concourse/main/deploy/s3-secret-access-key value=CHANGEME
```

And we re-trigger:

```
fly -t vm trigger-job -j deploy/deploy -w
started deploy/deploy #4

initializing
running sh -c set -o errexit
touch out/artifact-$(date +'%Y%m%d%H%M%S')

0 B / ? [----=]0
succeeded
```

We now (this is important!) remove the secret from Vault:

```
vault kv delete /concourse/main/deploy/s3-secret-access-key
Success! Data deleted (if it existed) at: concourse/main/deploy/s3-secret-access-key
```

And re-run again the deploy:

```
fly -t vm trigger-job -j deploy/deploy -w
started deploy/deploy #5

initializing
running sh -c set -o errexit
touch out/artifact-$(date +'%Y%m%d%H%M%S')

Expected to find variables: s3-secret-access-key
errored
```

:-)

So yes, it is a workaround and makes the secret accessible, for a short time window, to anybody who can login to Concourse team main AND replaces the pipeline named `deploy` with another that leaks the secret (so this is a malicious behavior).

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
